### DISCUSSION TITLE
Add parent info to callable Discriminator

### DISCUSSION THREAD

**Discussion by batzkass** [pydantic/pydantic#8149] [Feature Request]
Hi,
I just discovered the new callable Discriminators feature, that's great and it will help me in the future. It could be even more handy if there was a way to access parent instance data into the discriminator function. This would be helpful when the parent itself has a property that drives another property Union choice. Below is a basic example : an `Event` can be of category `"Music"` or `"Birthday"`. Depending on this category, the `Event` `guest_star` may be of type `Singer` or `Clown`. I'm looking for a way to drive the Union discriminator choice into the `Discriminator` function from the `Event` data, in this case the value of the `category` field.

```py
from typing import Literal, Annotated, Union, Any
from pydantic import BaseModel, Field, Tag, Discriminator

class Person(BaseModel):
    rating: int = Field(ge=0, le=10)

class Singer(Person):
    style: Literal["Jazz", "Rock"]

class Clown(Person):
    really_funny: bool

def discriminate_guest_star(v: Any):
    # Here I have no way to check the Event category field: "Music" or "Birthday" ?
    # It would be nice to be able to analyze Event data to return "singer" or "clown"
    return 'singer'

class Event(BaseModel):
    category: Literal["Music", "Birthday"]
    guest_star : Union[Annotated[Singer, Tag('singer')], Annotated[Clown, Tag('clown')]] = Field(discriminator=Discriminator(discriminate_guest_star))
```

To resolve this I currently use models inheritance, but it is more a kind of workaround. It makes cumbersome code : `Event` have to be subclassed and explicitly described for every `category` value, and wrapped into a kind of `EventContainer` with a discriminator based on the the `category` value. Also, for more complex cases this workaround would be almost impossible: imagine that you have multiple driving properties on the `Event`, this would make numerous subclasses...

Final note: I'm pretty new to pydantic, so there is a chance that I missed something and that another way of writing the above model exists. In this case don't hesitate to give me a hint about that.

Thanks !

**Comment by CaueP** [comment]
I have a similar question, and since there are no answers it's probably a feature request.

In my case, I don't particularly care about a discriminator function implementation, I would be ok to use a parent class field to identify what is the type of a `Union` field.

Below is an example code using the existing pet example from the docs:

```python

from typing import Union
from typing_extensions import Annotated

from pydantic import BaseModel, Discriminator, Field, Tag, ValidationError


class Cat(BaseModel):
    meows: int


class Dog(BaseModel):
    barks: float


class Lizard(BaseModel):
    scales: bool


class Model(BaseModel):
    pet_type: str    #  The attribute used on the discriminator is in the parent model
    pet: Annotated[Union[
            Annotated[Cat, Tag("cat")],
            Annotated[Dog, Tag("dog")],
            Annotated[Dog, Tag("lizard")]
        ],
        Discriminator("self.pet_type")   # Is there a way to use a parent attribute as a discriminator?
    ]
    n: int


print(Model(pet_type= 'dog', pet={'barks': 3.14}, n=1))
#> pet_type='dog' pet=Dog(barks=3.14) n=1

# The validation error would only show the missing attributes from Dog model
try:
    Model('pet_type': 'dog', pet={}, n=1)
except ValidationError as e:
    print(e)
    """
    1 validation error for Model
    pet.dog.barks
      Field required [type=missing, input_value={}, input_type=dict]
    """
```

Does anyone know if this is possible or have an alternative?

**Reply by mdomke** [reply]
I'm having the very same problem. I think it is also quite natural to have a sub-model whose shape is determined by a type discriminator on the parent model.

**Reply by rosscdh-tpg** [reply]
+1 same issue here

need to discrim on a childs parent field...

**Reply by jsouzadasilva** [reply]
+1
